store/copr/coprocessor.go

```diff
@@ -276,6 +313,11 @@ func buildCopTasks(bo *Backoffer, cache *RegionCache, ranges *KeyRanges, req *kv
 
 	tasks := make([]*copTask, 0, len(locs))
 	origRangeIdx := 0
+	taskID := uint64(0)
+	var store2Idx map[uint64]int
+	if req.StoreBatchSize > 0 {
+		store2Idx = make(map[uint64]int, 16)
```

Comment:
After the `cache.SplitKeyRangesByBuckets(bo, ranges)`, the ranges are split by the ordered region range, and the ranges within each region are also ordered. For example
```
Region 1            Region 2               Region 3
[1, 2], [3, 4]     [5, 10], [15, 20]        [21, 25]
task1   task2      task3  task4             task5
```
So if the `KeepOrder` is required, I think the batch processing could still work. The difference is that if the order is required, the coprocessor client could not response to the caller if `task5` has finished while `task2` does not.

@sticnarf @you06 
What do you think? Please correct me if I missed anything.

Reply:
I made a tiny change to the example.

```
Region 1   Region 2     Region 3               Region 4
[1, 2]     [3, 4]       [5, 10], [15, 20]      [21, 25]
task1      task2        task3  task4           task5

```

Suppose region1, region 2 and region 4 are located in store1, and region 3 is located in store2, there are two batch methods:

1. [task1, task2, task5], [task3, task4]

In this way, we archive the maximum batch size, and task5 should wait until task4 is received.

2. [task1, task2], [task3, task4], [task5]

In this way, we don't need to reorder the responses.

Reply:
This largely reduces the effect of batching. It's possible that a batch involves hundreds of regions in one store. It's very common that region ranges intersects between stores.

Instead, I think we should store the range or the order index of the response and sort them after receiving all of them. This can be done in the next iterations.

Reply:
Checking about the ordering related work in the next iteration is fine to me, by now we could just disable batching if order is required to make it simple.